### PR TITLE
feat: add support for computing coverage per line

### PR DESCRIPTION
**Pull request by martinezmatias**
The main change introduced by this PR is the ability to pass a new class for computing the coverage. 
Previously, the class was hardcoded (see for example [here](https://github.com/STAMP-project/test-runner/blob/master/src/main/java/eu/stamp_project/testrunner/listener/junit4/JUnit4Coverage.java#L19)).
That previous coverage engine did not store the coverage per line, it  summarizes the results.
Now, I change the interfaces of such classes to be able to pass a new coverage engine. By default, it uses that one used in the previous version. This change able us to pass other implementations that computes coverage (e.g. [coverage by line](https://github.com/STAMP-project/test-runner/compare/master...martinezmatias:extension_fl?expand=1#diff-cd96cafcf009a661f3a5830e0bb266daR24) )
In addition, we added a new method in the coverage interface to retrieve more detailed information (e.g., the lines that are covered) (Previously it only returned the number of covered lines, but not which ones were covered)

### CODE REVIEW

File: src/main/java/eu/stamp_project/testrunner/runner/coverage/JacocoRunner.java [hunk 1]
```diff
@@ -127,6 +144,16 @@ public Coverage run(String classesDirectory,
     protected abstract Coverage executeTest(String[] testClassNames,
                                             String[] testMethodNames,
                                             List<String> blackList);
+    
+    
+    
+    
+    protected Coverage executeTest(String[] testClassNames,
```
Comment (monperrus) [inline src/main/java/eu/stamp_project/testrunner/runner/coverage/JacocoRunner.java:151]: coverage seems to be the output object, what information does it contain?

**Review by monperrus — COMMENTED** [review]
Hi Matias,


That's really cool!! Overall the changes are good and backward compatible.

My main comment is that having a method `executeTest` which takes as input a `Coverage` is strange, because the Coverage is more the output. 

What about using the builder pattern instead:

```
Runner j = new JacocoRunner(new XYZCOverageBuilder());
j.executeTest("Test","test", []) // calls XYZCOverageBuilder.buildCoverage internally?
```

**Comment by martinezmatias** [comment]
Hi Martin

> My main comment is that having a method executeTest which takes as input a Coverage is strange, because the Coverage is more the output.

It fully agree with you. Coverage should not be the input.
I passed as input parameter because that Coverage class has one method we call to compute the coverage:

```

public void collectData(ExecutionDataStore executionData, String classesDirectory);

```
[Coverage](https://github.com/STAMP-project/test-runner/blob/master/src/main/java/eu/stamp_project/testrunner/listener/Coverage.java#L27) 

This issue arises because the Coverage declares both the method to computes the coverage (e.g., `collectData`)  and output information (e.g. `getInstructionsCovered()`). 

I would propose to split in two interfaces. So, one would be the input of the test execution and another the output.

File: src/main/java/eu/stamp_project/testrunner/runner/coverage/JacocoRunner.java (hunk 1 above)
Comment (martinezmatias) [reply src/main/java/eu/stamp_project/testrunner/runner/coverage/JacocoRunner.java:151]: Coverage contains, for example, a method that returns the number of covered lines, but, at the same time, it declares a method for collecting data. Due to that method, I passed as input.

**Comment by monperrus** [comment]
> I would propose to split in two interfaces. So, one would be the input of the test execution and another the output.

yes!

**Comment by martinezmatias** [comment]
This new commit introduces the following changes:

* New interface `TestCoveredResult` that combines TestResults and Coverage: it represents the results of a test that was instrumented for computing coverage: https://github.com/STAMP-project/test-runner/pull/80/commits/5423535c010de297f8fc874e692929bc24cb1f6b#diff-7da9a90a2ff891191ddeeb4667ca54f6R6

* Changes in the interface executeTest: before it returned a Coverage object and then subsequent castings where done (e.g, to test results). Now, the executionTest returns a `TestCoveredResult`. E.g https://github.com/STAMP-project/test-runner/pull/80/commits/5423535c010de297f8fc874e692929bc24cb1f6b#diff-a388d06d8d9eac06d0133aa6463d9310R66


* Decoupling Coverage class. The computation of coverage is done by interface `CoverageCollector` https://github.com/STAMP-project/test-runner/pull/80/commits/5423535c010de297f8fc874e692929bc24cb1f6b#diff-61a6bdf8bb6d700675730334dbd59caeR10

* Now, the interface Coverage stores the result of the coverage.

**Comment by monperrus** [comment]
* commented
* refactored
* updated to newer Jacoco 0.8.3
* improved test harness to better run locally and in IntelliJ

**Comment by monperrus** [comment]
I will merge later today or tomorrow

**Comment by monperrus** [comment]
Thanks a lot @martinezmatias

**Comment by monperrus** [comment]
Now the fun continues on https://github.com/SpoonLabs/flacoco/pull/2

**Pull request merged**